### ISSUE TITLE
IMappingExpression.BeforeMap((x,y) => Mapper.Map()) stack overflow exception in 4.1.1

### ISSUE THREAD

**Issue by sgwatgit** [AutoMapper/AutoMapper#963]
I found an example of how to have AutoMapper use the mappings of a base class of the destination object so they wouldn't have to be re-defined for the inheriting object.

```
public static void InheritMappingFromDestinationBaseType<TSource, TDestination>(this IMappingExpression<TSource, TDestination> mappingExpression)
        {
            var sourceType = typeof(TSource);
            var desctinationType = typeof(TDestination);
            var sourceParentType = sourceType;
            var destinationParentType = desctinationType.BaseType;

            mappingExpression
                .BeforeMap((x, y) => Mapper.Map(x, y, sourceParentType, destinationParentType))
                .ForAllMembers(x => x.Condition(r => NotAlreadyMapped(sourceParentType, destinationParentType, r)));
        }
```

This worked fine in `3.x` but after upgrade to `4.1.1`, the same maps/code base now throws a `StackOverflowException`.

My maps are defined as follows

```
public class UserToUserAggregateDTOMapper : Profile
    {
        protected override void Configure()
        {
            base.Configure();

            CreateMap<User, UserAggregateDTO>()
                .ForAllMembers(e => e.Ignore());

            CreateMap<User, UserAggregateDTO>()
                .ForMember(d => d.Name, e => e.MapFrom(i => i.UserName))
                .ForMember(d => d.Friends, e => e.MapFrom(i => i.UserFriends))
                .InheritMappingFromDestinationBaseType();
        }
    }
```
In this example, the `UserAggregateDTO` inherits from a `UserDTO` class that has some other properties on it which are mapped in another mapping profile.

I'm not sure what is causing this exception to be thrown, it doesn't happen with all maps. The maps that don't use this extension method don't have any issues. The exception is thrown at the line `.BeforeMap((x, y) => Mapper.Map(x, y, sourceParentType, destinationParentType))`

Is anyone aware of any changes between 4.x and 3.x that I need to be aware of in my configuration or the definition of this extension method?

**Comment by lbargaoanu** [comment]
This should work with IncludeBase (or Include). It doesn't because of a bug fixed by #958.

**Comment by sgwatgit** [comment]
@lbargaoanu Thanks, I believe this method was not working with 4.0.0 was first released but I see that a fix for it is now included.

I tested `IncludeBase` and removed my extension methods and I'm not running on version 4.4.1.
